compiler/rustc_builtin_macros/src/deriving/debug.rs

```diff
@@ -94,12 +94,17 @@ fn show_substructure(cx: &mut ExtCtxt<'_>, span: Span, substr: &Substructure<'_>
         args.extend([fmt, name]);
         for i in 0..fields.len() {
             let field = &fields[i];
-            if is_struct {
-                let name = cx.expr_str(field.span, field.name.unwrap().name);
+            if let Some(name) = field.name {
+                let name = cx.expr_str(field.span, name.name);
                 args.push(name);
             }
-            // Use an extra indirection to make sure this works for unsized types.
-            let field = cx.expr_addr_of(field.span, field.self_expr.clone());
+
+            let field = if i < fields.len() - 1 {
```

Comment:
This if/else is now complicated enough that it might be worth factoring out, e.g. into an inner function, to avoid the duplication with the code below.